### ISSUE TITLE
reproduction of the monodepth2 with resnet18

### ISSUE THREAD

**Issue by MingYang-buaa** [TRI-ML/packnet-sfm#40]
Hey! Thanks for your wonderful work! 

 Now i'm trying to reproduce the result of monodepth2 with resnet18 backbone. I used the train_kitti.yaml to train about 50 epochs. But the result is not good with loss around about 0.075 but absrel is only 0.125. I think it is may overfitting with the imperfect hyperparameters. So can you share the .yaml file for training monodepth2?

**Comment by MingYang-buaa** [comment]
And the infer result is also not good as that.
![test](https://user-images.githubusercontent.com/67720091/86308231-34a4b500-bc4b-11ea-9cf1-56f6129c7f79.png)

**Comment by VitorGuizilini-TRI** [comment]
We routinely get < 0.12 abs_rel with monodepth2, just by changing from PackNet to DepthResNet. Some possible reasons:

- Are you using ImageNet pretrained weights (version='18pt') for both DepthResNet and PoseResNet? Monodepth2 really needs ImageNet pretraining, otherwise results are much worse (another reason we like PackNet, it doesn't need supervised pretrained weights). 
- Try decreasing the learning rate a bit (1e-4), and increasing batch size (8).

Let me know how it goes.

**Comment by MingYang-buaa** [comment]
1. yes, i used the ImageNet pretrained weights for both DepthResNet and PoseResNet.
2. the learing rate is 2e-4 for both depth net and pose net. Then something i changed is step_size(from 30 to 15), batch_size(from 4 to 12). And i'm training on 4 GPUs(2080ti) rather than 8GPUS(V100).
that's the problem?

**Comment by MingYang-buaa** [comment]
By the way. Did you try to train monodepth2 with VelSupModel? If true, how about the result?

**Comment by VitorGuizilini-TRI** [comment]
Try bringing your learning rate down to 1e-4, just to make sure, and I will start a training session here with your configuration as soon as possible, again just to make sure. We've tried monodepth2 with VelSupModel, usually we achieve metrically accurate results similar to SelfSupModel.

**Comment by MingYang-buaa** [comment]
Thanks for your so quick reply! I also will try to train the SelfSupModel and SelfSupModel with the lr as 1e-4. Looking forward to your good news.

**Comment by VitorGuizilini-TRI** [comment]
Hi, results seem to be OK here (attached image from my latest run with resnet18 backbone), so I am not sure what could be the problem from your end. Regardless, it would be interesting to check if our losses match exactly those from monodepth2, as we are using some of their tricks at training time.  

![val-KITTI_raw-eigen_test_files-velodyne-544-inv_depth_49_afb5bd80](https://user-images.githubusercontent.com/58576956/86371345-9eb25f80-bc35-11ea-8538-c936c1e9c812.png)

**Comment by MingYang-buaa** [comment]
All right. It's confusing that i also test to train with lr in 1e-4 but the abs_rel only reach to 0.120. 
Can you shared the yaml file?  I would appreciate it.
Also the tricks you mentioned above is that the matching resolution ans camera intrinsic params? I only noticed these :)

**Comment by VitorGuizilini-TRI** [comment]
I will share the .yaml file with you shortly, but I haven't changed much from the repository one.
About the tricks, there is also auto-masking and minimum reprojection loss. 
Was your 0.120 the minimum value or the end value? I noticed that monodepth2 tends to degrade a little if you train for too long. 
Training in a no-docker environment should be fine, but perhaps there are some differences in library versions that are making an impact.

**Comment by MingYang-buaa** [comment]
yeap. I had trained about 30 epochs, but about the last 10 epochs of asb_rel values float around 0.120. So i stoped it.
And i made a mistake that the pic I posted for the first time is trained with VelSupModel. But this is also strange about that.
Have you encountered this wrong result?
Maybe I have to train for a few more rounds to find out why.

**Comment by MingYang-buaa** [comment]
@VitorGuizilini-TRI  Hi. Can you share the monodepth2.yaml now? 
Other than that, the _depth regularization weight_ you mentioned in paper is included in the code now? I cannot find it. 
And the velocity-scaling weight is little different that 0.05 in paper but 0.1 in default_config. Does it meant that 0.1 is better than 0.05 in experirent?

**Comment by pjckoch** [comment]
Hi @MingYang-buaa ,

could you give any insights on whether you were able to resolve this issue? I encountered the same problem with the VelSupModel. For me the solution was to first train self-supervised and introduce the velocity supervision after a few epochs. If training with velocity supervision from the beginning, the output would look exactly like yours. I observed that the error lies in a wrongly estimated pose estimation: when looking at the warped reference images, the one that is generated from **t-1** looks ok, but the one from **t+1** does not. In fact, the one from **t+1** seems more like it was generating a warped **t+2** frame, instead of a warped **t** frame. That implies that the pose transformation is always estimated to be forward, never backwards.

Please let me know if you have any further insights on that.

Thank you. Best regards,
Patrick
